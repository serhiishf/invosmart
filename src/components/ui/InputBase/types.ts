import { InputHTMLAttributes } from 'react';

type InputPadding = 'default' | 'moderate' | 'minimal' | 'none';

export interface InputBaseProps extends InputHTMLAttributes<HTMLInputElement> {
  paddingLeft?: InputPadding;
  paddingRight?: InputPadding;
  isError?: boolean;
  /**
   * The isReadOnlyMode prop is used instead readonly attribute to keep inputs focusable and part of the tab order on iOS Safari, where the readonly attribute would otherwise remove them.
   */
  isReadOnlyMode?: boolean;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}
