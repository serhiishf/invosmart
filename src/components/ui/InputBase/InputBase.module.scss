.baseInput {
  width: 100%;
  height: 100%;
  outline: none;
  box-shadow: none;
  font-family: $font-family-primary;
  font-size: $font-size-base;
  font-weight: 500;
  color: $color-field-text;
  background-color: transparent;
  border: none;
  padding: 0 26px;

  &::placeholder {
    color: $color-field-placeholder;
  }

  &:focus {
    border: none;
    outline: none;
    box-shadow: none;
  }
}
