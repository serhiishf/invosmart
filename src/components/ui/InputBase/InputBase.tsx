import classNames from 'classnames';
import { forwardRef, useState, useEffect } from 'react';
import styles from './InputBase.module.scss';
import { InputBaseProps } from './types';

const InputBase = forwardRef<HTMLInputElement, InputBaseProps>(
  (
    {
      paddingLeft = 'default',
      paddingRight = 'default',
      isReadOnlyMode = false,
      value,
      onChange,
      ...rest
    }: InputBaseProps,
    ref
  ) => {
    const [inputValue, setInputValue] = useState(value ?? '');

    useEffect(() => {
      setInputValue(value ?? '');
    }, [value]);

    const handleChangeIfAllowed = (event: React.ChangeEvent<HTMLInputElement>) => {
      if (isReadOnlyMode) return;
      setInputValue(event.target.value);
      if (!onChange) return;
      onChange(event);
    };

    return (
      <input
        className={classNames(
          styles.inputBase,
          styles[`inputBase--paddingLeft-${paddingLeft}`],
          styles[`inputBase--paddingRight-${paddingRight}`],
          isReadOnlyMode && styles['inputBase--readOnlyMode']
        )}
        onChange={handleChangeIfAllowed}
        ref={ref}
        value={inputValue}
        {...rest}
      />
    );
  }
);

export default InputBase;
