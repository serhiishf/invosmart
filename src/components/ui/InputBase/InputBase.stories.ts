import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import InputBase from './InputBase';
import { PaddingInput } from 'constants/theme';

const meta = {
  title: 'components/UI/InputBase',
  component: InputBase,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    paddingLeft: { control: 'radio', options: PaddingInput },
    paddingRight: { control: 'radio', options: PaddingInput },
  },
  args: {
    onChange: fn(),
  },
} satisfies Meta<typeof InputBase>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {
  args: {
    placeholder: 'Placeholder',
    value: 'Value',
    isReadOnlyMode: false,
  },
};
