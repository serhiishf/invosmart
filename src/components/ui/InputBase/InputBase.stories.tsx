import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import InputBase from './InputBase';

const meta = {
  title: 'components/UI/InputBase',
  component: InputBase,
  decorators: [
    (Story) => (
      <div style={{ border: '1px dashed pink' }}>
        <Story />
      </div>
    ),
  ],
  tags: ['autodocs'],
  args: {
    onChange: fn(),
  },
} satisfies Meta<typeof InputBase>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {};

export const InitialValue: Story = {
  args: {
    placeholder: 'Placeholder',
    value: 'Value',
  },
};

export const Placeholder: Story = {
  args: {
    placeholder: 'Placeholder',
  },
};

export const ReadonlyMode: Story = {
  args: {
    value: 'Readonly mode',
    isReadOnlyMode: true,
  },
};

export const Disabled: Story = {
  args: {
    value: 'Disabled',
    disabled: true,
  },
};
