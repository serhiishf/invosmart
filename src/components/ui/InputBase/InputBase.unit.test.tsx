import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import InputBase from './InputBase';

describe('InputBase', () => {
  describe('default Initialization', () => {
    it('should render correctly with default props', () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      expect(input).toBeInTheDocument();
      expect(input).not.toHaveValue();
    });

    it('should not have value and placeholder attribute', () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      expect(input).not.toHaveValue();
      expect(input).not.toHaveAttribute('placeholder');
    });

    it('should update the value on user input', async () => {
      render(<InputBase />);
      const input = screen.getByRole('textbox');
      await userEvent.type(input, 'typed text');
      expect(input).toHaveValue('typed text');
    });

    it('should match the snapshot with default props', () => {
      const { asFragment } = render(<InputBase />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('initialization with props', () => {
    it('should render empty input with placeholder', () => {
      render(<InputBase placeholder="Some placeholder" />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveAttribute('placeholder', 'Some placeholder');
    });

    it('should render the input correctly with an initial value', () => {
      render(<InputBase value="initial value" />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveValue('initial value');
    });

    it('should mark the input as aria-readonly and prevent typing when isReadOnlyMode is true', async () => {
      render(<InputBase isReadOnlyMode={true} />);
      const input = screen.getByRole('textbox');

      // Ensure the input is marked as read-only for accessibility
      expect(input).toHaveAttribute('aria-readonly', 'true');

      // Attempt to type into the input and check that the input value is unchanged
      await userEvent.type(input, 'typed text');
      expect(input).toHaveValue('');
    });

    it('should retain the initial value and prevent typing when isReadOnlyMode is true', async () => {
      render(<InputBase isReadOnlyMode={true} value="some text" />);
      const input = screen.getByRole('textbox');

      // Ensure the input is marked as read-only for accessibility
      expect(input).toHaveAttribute('aria-readonly', 'true');

      // Attempt to type into the input and check that the initial value is unchanged
      await userEvent.type(input, 'typed text');
      expect(input).toHaveValue('some text');
    });

    it('should have attribute aria-invalid="true" when isError=true', () => {
      render(<InputBase isReadOnlyMode={true} value="some text" isError={true} />);
      const input = screen.getByRole('textbox');

      expect(input).toHaveAttribute('aria-invalid', 'true');
    });

    it('should render disabled input and prevent any typing', async () => {
      render(<InputBase disabled={true} />);
      const input = screen.getByRole('textbox');
      expect(input).toBeDisabled();
      await userEvent.type(input, 'Some text');
      expect(input).not.toHaveValue();
    });

    it('should match the snapshot when isReadOnlyMode is true', () => {
      const { asFragment } = render(<InputBase isReadOnlyMode={true} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
