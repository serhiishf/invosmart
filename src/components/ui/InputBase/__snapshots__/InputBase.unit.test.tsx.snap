// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputBase > Default Initialization > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <input
    class="_inputBase_6ca9ba _inputBase--paddingLeft-default_6ca9ba _inputBase--paddingRight-default_6ca9ba"
    value=""
  />
</DocumentFragment>
`;

exports[`InputBase > initialization with props > should match the snapshot when isReadOnlyMode is enabled 1`] = `
<DocumentFragment>
  <input
    class="_inputBase_6ca9ba _inputBase--paddingLeft-default_6ca9ba _inputBase--paddingRight-default_6ca9ba _inputBase--readOnlyMode_6ca9ba"
    value=""
  />
</DocumentFragment>
`;
